vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go

```diff
@@ -0,0 +1,884 @@
+	for i, si := range srcSet[len(curSet):] {
+		sm := si.(map[string]interface{})
+		r := NewNetworkInterfaceSubresource(c, d, sm, nil, i+len(curSet))
+		dspec, err := r.Delete(l)
+		if err != nil {
+			return nil, nil, fmt.Errorf("%s: %s", r.Addr(), err)
+		}
+		l = applyDeviceChange(l, dspec)
+		spec = append(spec, dspec...)
+	}
+
+	log.Printf("[DEBUG] NetworkInterfacePostCloneOperation: Post-clone final resource list: %s", subresourceListString(updates))
+	// We are now done! Return the updated device list and config spec. Save updates as well.
+	if err := d.Set(subresourceTypeNetworkInterface, updates); err != nil {
+		return nil, nil, err
+	}
+	log.Printf("[DEBUG] NetworkInterfacePostCloneOperation: Device list at end of operation: %s", DeviceListString(l))
+	log.Printf("[DEBUG] NetworkInterfacePostCloneOperation: Device config operations from post-clone: %s", DeviceChangeString(spec))
+	log.Printf("[DEBUG] NetworkInterfacePostCloneOperation: Operation complete, returning updated spec")
+	return l, spec, nil
+}
+
+// ReadNetworkInterfaceTypes returns a list of network interface types. This is used
+// in the VM data source to discover the types of the NIC drivers on the
+// virtual machine. The list is sorted by the order that they would be added in
+// if a clone were to be done.
+func ReadNetworkInterfaceTypes(l object.VirtualDeviceList) ([]string, error) {
+	log.Printf("[DEBUG] ReadNetworkInterfaceTypes: Fetching interface types")
+	devices := l.Select(func(device types.BaseVirtualDevice) bool {
+		if _, ok := device.(types.BaseVirtualEthernetCard); ok {
+			return true
+		}
+		return false
+	})
+	log.Printf("[DEBUG] ReadNetworkInterfaceTypes: Network devices located: %s", DeviceListString(devices))
+	// Sort the device list, in case it's not sorted already.
+	devSort := virtualDeviceListSorter{
+		Sort:       devices,
+		DeviceList: l,
+	}
+	sort.Sort(devSort)
+	devices = devSort.Sort
+	log.Printf("[DEBUG] ReadNetworkInterfaceTypes: Network devices order after sort: %s", DeviceListString(devices))
+	var out []string
+	for _, device := range devices {
+		out = append(out, virtualEthernetCardString(device.(types.BaseVirtualEthernetCard)))
+	}
+	log.Printf("[DEBUG] ReadNetworkInterfaceTypes: Network types returned: %+v", out)
+	return out, nil
+}
+
+// baseVirtualEthernetCardToBaseVirtualDevice converts a
+// BaseVirtualEthernetCard value into a BaseVirtualDevice.
+func baseVirtualEthernetCardToBaseVirtualDevice(v types.BaseVirtualEthernetCard) types.BaseVirtualDevice {
+	switch t := v.(type) {
+	case *types.VirtualE1000:
+		return types.BaseVirtualDevice(t)
+	case *types.VirtualE1000e:
+		return types.BaseVirtualDevice(t)
+	case *types.VirtualPCNet32:
+		return types.BaseVirtualDevice(t)
+	case *types.VirtualSriovEthernetCard:
+		return types.BaseVirtualDevice(t)
+	case *types.VirtualVmxnet2:
+		return types.BaseVirtualDevice(t)
+	case *types.VirtualVmxnet3:
```

Comment:
This actually does not work:

* This switch would pass everything except for the final case as a no-op
* You cannot `fallthrough` on a type switch - I've tried :wink: 

However, let me see if this can be done better. I have a feeling it can.

Reply:
New function:

```
func baseVirtualDeviceToBaseVirtualEthernetCard(v types.BaseVirtualDevice) (types.BaseVirtualEthernetCard, error) {
  if bve, ok := v.(types.BaseVirtualEthernetCard); ok {
    return bve, nil
  }
  return nil, fmt.Errorf("device is not a network device (%T)", v)
}
```

:laughing:

I think at some point in time I tried to assign a `BaseVirtualDevice` value to a `BaseVirtualEthernetCard` value directly, and of course that fails due to one interface not implementing the other, so I arrived at the conclusion for some reason that this kind of assignment wasn't possible period, even with type assertion. Anyway, this is cleaned up now and could possibly be cleaned up further, but this will serve the purpose and has the extra benefit of failing gracefully on assertion failure.